### PR TITLE
planner: Modify the judgment condition of rule_generate_column_substitute.

### LINKED ISSUE #16782: The intercepted value will use the `rule_generate_column_substitute` incorrectly.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```sql
CREATE TABLE t0(c0 tinyint AS (c1) UNIQUE, c1 INT);
set sql_mode="";
INSERT INTO t0(c1) VALUES (1000);
select * from t0;
```

### 2. What did you expect to see?

```sql
+------+-------+
| c0   | c1    |
+------+-------+
|  127 |  1000 |
+------+-------+
1 row in set (0.00 sec)
```

### 3. What did you see instead?

```sql
+------+------+
| c0   | c1   |
+------+------+
|  127 |  127 |
+------+------+
1 row in set (0.00 sec)
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
```sql
+----------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------+
| tidb_version()

                                                            |
+----------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-341-g35a801bc9
Git Commit Hash: 35a801bc9ba21a8013abaa2c9143bab7f083ebc5
Git Branch: master
UTC Build Time: 2020-04-24 02:57:37
GoVersion: go1.14.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+----------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by LittleFall**
### What problem does this PR solve?

Issue Number: close #16782 

Problem Summary:

[pull/16316](https://github.com/pingcap/tidb/pull/16316) modified the rule of rule_generate_column_substitute. After that, there are still some problems. For example, [issues/16782](https://github.com/pingcap/tidb/issues/16782). In this PR, the judgment rules are adjusted more strictly and appropriately.

### What is changed and how it works?

In [parser/pull/841](https://github.com/pingcap/parser/pull/841), the Equal function was updated. 
In this pr, we directly use the Equal function to check whether the types are equal.

In current:

Double and float cannot be substituted with each other, because different storage space will cause panic.

The doubles(floats are the same) defined by default can be substituted with each other, and can also be substituted with the calculated double. There Decimal is both -1, only have different Flen. 

The double defined with `double(p)` is the same as above, only Flen is different.

The double defined by `double(m, d)` you will get a Decimal value other than -1. When Decimal is specified, doubles can only be substituted if Decimal and Flen are all equal.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

Enhance the security of rule_generate_column_substitute.

### CODE REVIEW

File: planner/core/rule_generate_column_substitute.go [hunk 1]
```diff
@@ -45,25 +45,15 @@ func (gc *gcSubstituter) optimize(ctx context.Context, lp LogicalPlan) (LogicalP
...
-	}
-	return int(colType.Tp)
-}
-
-// We do not check whether flen is the same.
-// Because different flen will not cause data truncation.
-// See https://github.com/pingcap/tidb/pull/16316.
+// When Tp is an integer or real type, we don't check whether Flen is equal.
+// Because their Flen is useless for them in TiDB now, and may be unequal for some reasons.
+// For example, Tp for default real is 22, and Tp for sum of real is -1.
```
Comment (Deardrops) [inline planner/core/rule_generate_column_substitute.go:50]: Any test case cover this?

File: planner/core/rule_generate_column_substitute.go [hunk 2]
```diff
@@ -45,25 +45,15 @@ func (gc *gcSubstituter) optimize(ctx context.Context, lp LogicalPlan) (LogicalP
...
 func checkSubstitutability(c1, c2 *types.FieldType) bool {
-	partialEqual := c1.Decimal == c2.Decimal &&
+	partialEqual := c1.Tp == c2.Tp &&
```
Comment (wjhuang2016) [inline planner/core/rule_generate_column_substitute.go:52]: Please modify the function `equal` directly.

**Pull request merged**